JASP-Engine/JASP/R/multinomialtest.R

```diff
@@ -15,546 +15,447 @@
   #
   # Return:
-  #   Object (chisquare test)
-
+  #   No return, stores jaspResults Object (chisquare test)
+ 
+  # Take results from state if possible
+  if (!is.null(jaspResults[["stateChisqResults"]])) 
+    return(jaspResults[["stateChisqResults"]]$object)
+  
   chisqResults <- NULL
-
-  if (perform == "run" && !is.null(fact)) {
+  fact <- options$factor
+  
+  if(fact != ""){
     # first determine the hypotheses
     factorVariable <- dataset[[.v(fact)]]
+    factorVariable <- factorVariable[!is.na(factorVariable)]
     factorVariable <- as.factor(factorVariable)
     nlev <- nlevels(factorVariable)
-
+    
     if (options$counts != "") {
       counts <- dataset[[.v(options$counts)]]
       # omit count entries for which factor variable is NA
       counts <- counts[!is.na(factorVariable)]
-      # check for invalid counts
-      .checkCountsMultinomial(counts, nlev)
-      dataTable        <- counts
-      names(dataTable) <- levels(factorVariable)
-    } else {
-      dataTable <- table(factorVariable)
-    }
-
+      c <- dataset[[.v(options$counts)]]
+      factorVariable <- factor(rep(factorVariable, c), 
+                               levels = levels(factorVariable))
+    } 
+    
+    dataTable <- table(factorVariable)
+    
     hyps <- .multinomialHypotheses(dataset, options, nlev)
-
+    
     # create a named list with as values the chi-square result objects
     chisqResults <- lapply(hyps, function(h) {
       # catch warning message and append to object if necessary
-      csr <- NULL
-      warn <- NULL
-      csr <- withCallingHandlers(
-        chisq.test(x = dataTable, p = h, rescale.p = TRUE,
-                   simulate.p.value = options$simulatepval),
-        warning = function(w) {
-         warn <<- w$message
-        }
+      csr  <- warn <- NULL
+      # need to improve this try statement
+      csr  <- withCallingHandlers(
+                chisq.test(x = dataTable, p = h, rescale.p = TRUE,
+                           simulate.p.value = options$simulatepval),
+                warning = function(w) warn <<- w$message
       )
       csr[["warn"]] <- warn
       return(csr)
     })
   }
+  # Save results to state
+  jaspResults[["stateChisqResults"]] <- createJaspState(chisqResults)
+  jaspResults[["stateChisqResults"]]$dependOn(c("factor", "counts", "hypothesis", 
+                                                "exProbVar", "tableWidget", "simulatepval"))
+}
 
-  return(chisqResults)  # return the out object
+# Filling Tables ----
+.multinomMainTableFill <- function(jaspResults, dataset, options) {
+  # Turn chi-square object into jaspResults 
+  # object and creates results for descripties table
+  #
+  # Args:
+  #   jaspResults:
+  #   dataset: input dataset
+  #   options: user options
+  #   chisqResults:
+  #   
+  # Return:
+  #   No return, stores jaspResults Object (main table results)
+  
+  results <- list()
+
+  # Compute Results
+  .chisquareTest(jaspResults, dataset, options)
```

Comment:
```suggestion
  chisqResults <- .chisquareTest(jaspResults, dataset, options)
```
why not have `.chisquareTest` check the state and return the object if it exists, or otherwise compute and return the object? Now the next line is getting the object from state, although it could have been returned.